### PR TITLE
First version of the serve and store service

### PR DESCRIPTION
**Pull request by rmeissn**
I've transformed the service to one that needs a MongoDB and can handle uploaded pictures, as well as is backward compatible.

Some test cases are missing, but might need some time to implement. My best guess is ~1 week.

Please review carefully and also review what is stored in the MongoDB and if that matches the proposed model - https://slidewiki.atlassian.net/wiki/display/SWIK/Media?focusedCommentId=31490055#comment-31490055
I'm open for ideas and proposals.

Open Questions:
* Is it a good idea to store files under the hash of the original file (not transformed)?
* Is it a good idea to clear and convert metadata instead of storing it as a string?
* collisions of files (very unlikely) is not handled

Please only state if you think it's okay or not, because I want to review and merge it myself - merge only if you think it's okay.

### CODE REVIEW

File: application/configuration.js [hunk 1]
```diff
@@ -1,17 +1,51 @@
...
+      host = entrys[entrys.length - 1];
+      console.log('Found mongodb host. Using ' + host + ' as database host.');
+    }
+  }
+} catch (e) {
+  //Windows or no read rights (bad)
+}
+
 const co = require('./common');
-let port = 8080;
```
Comment (rmeissn) [inline application/configuration.js]: APPLICATION_Port is missing --> that's mandetory

**Comment by TBoonX** [comment]
Looks good! If you could test it with success then lets use it for the future!

Data model: Think about the attributes alt, caption and usage.

Questions:
I would store the files with an ObjectId as filename (plus ending) - with this there will be no conflicts. (For duplication detection you could save the filehash in the media data model)
If the most important information of the metadata are stored then it is OK to clean it and save it as key value pairs.

**Comment by rmeissn** [comment]
About the data model: the idea is to not reimplement parts of the EXIF or ID3Tag definitions in our model, but instead use the appended metadata object for these kind of information.
What do you mean by usage?

I'll have a look at the ObjectID idea....seems to resolve some of the problems I am facing.
EDIT: I have to disagree. When a user uploads a picture, I can only search for it in the databse by its hash sum. By using ObjectIDs instead of the hashs as picture names, conflicts are not solved, because the file with the same sum will be twice in the database. Or we just ignore duplicates (more HDD space + computing time for conversions) and save every uploaded picture, regardless whether it has been already uploaded or not. --> What's better? A conflict is still very, very (!) unlikely.

Well, I hope that not too much metadata is discarded. At least for my testing pics, it works pretty good.

**Comment by TBoonX** [comment]
Saving the parsable metadata is good. Thus no reimplementation of EXIF or so.
In Confluence the media data model has the attribute "usage" - please think about if this is necessary and/or it should be stored in a deck revision.

I think its not that unlikely that two or more users uploading the exact same picture.
I think its a bad practice to either link two media data instances to one file or notify the user that this picture is already in slidewiki.
Thus we have to life with duplication of files. Better 300MB disk usage more than duplication of files.

**Comment by rmeissn** [comment]
What is your opinion on duplicates, @kadevgraaf ? Save duplicates in each and any case as a new picture or just link to the already saved picture? I already implemented the second one.

**Comment by kadevgraaf** [comment]
Sorry for my late response.
W.r.t. duplicates; If you already implemented the second one (link to same picture) then I guess it is best to go for that. Especially if conflicts are very unlikely. It might be bad practice, but unless there is something that would cause conflicts it seems best at this moment w.r.t. planning.
To prevent conflicts we would for example have to take into account that images are stored in user directories, so if a users disables his/her account, not delete this directory. Also, users cannot delete images unless we can detect that another slide/user uses the image.

Looks good, especially the backward compatibility is nice 👍

**Comment by rmeissn** [comment]
Well, I haven't implemented deletion of pictures yet ... consider it as a feature :P
Okay, so I will execute a final test with Kurt today and if everything works, merge it.

**Comment by rmeissn** [comment]
Kurt discovered a few problems with the parsage of metadata object ... he tries to write a custom parser instead of using the yaml parser. Fallback would be to store it as an array of strings.

**Comment by TBoonX** [comment]
Just one thing doesn't work for me:
If you send a file which is bigger than 10MB the service crashes.

```
/home/kjunghanns/Documents/Slidewiki/file-service/application/controllers/handler.js:11
    if (request.payload.bytes <= 1) { //no payload
                       ^

TypeError: Cannot read property 'bytes' of null
    at storePicture (/home/kjunghanns/Documents/Slidewiki/file-service/application/controllers/handler.js:11:24)
    at Object.internals.handler (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/hapi/lib/handler.js:96:36)
    at request._protect.run (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/hapi/lib/handler.js:30:23)
    at internals.Protect.run (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/hapi/lib/protect.js:62:12)
    at exports.execute (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/hapi/lib/handler.js:24:22)
    at each (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/hapi/lib/request.js:380:16)
    at iterate (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/hapi/node_modules/items/lib/index.js:36:13)
    at done (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/hapi/node_modules/items/lib/index.js:28:25)
    at postValidate (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/hapi/lib/validation.js:68:20)
    at _validateWithOptions (/home/kjunghanns/Documents/Slidewiki/file-service/application/node_modules/joi/lib/any.js:604:20)

```

**Pull request merged**